ARCH-1253: logistration using login user

robrap opened the pull request:
use login_ajax for logistration

ARCH-1253

### Please consider the following when opening a pull request:

- Link to the relevant JIRA ticket(s) and tag any relevant team(s).
- Squash your changes down into one or more discrete commits.
  In each commit, include description that could help a developer
  several months from now.
- If running `make upgrade`, run _as close to the time of merging as possible_
  to avoid accidentally downgrading someone else's package.
  Put the output of `make upgrade` in its own separate commit,
  decoupled from other code changes.
- Aim for comprehensive test coverage, but remember that
  automated testing isn't a substitute for manual verification.
- Carefully consider naming, code organization, dependencies when adding new code.
  Code that is amenable to refactoring and improvement benefits all platform developers,
  especially given the size and scope of edx-platform.
  Consult existing Architectural Decision Records (ADRs),
  including those concerning the app(s) you are changing and
  [those concerning edx-platform as a whole](https://github.com/edx/edx-platform/tree/master/docs/decisions).

nasthagiri commented on lms/static/js/student_account/views/LoginView.js:
```diff
@@ -189,11 +189,18 @@
+                var msg;
+                // TODO: Add back the following failsafe line once tests are passing,
+                //   or a version of it that adds a default message.
+                // var msg = error.responseText;
```
Leftover commented code?

robrap replied:
Oops.  Updated.  I had a `Reviewer Note` (see above) that had explained, but I didn't realize it was Pending.

nasthagiri commented on lms/static/js/student_account/views/LoginView.js:
```diff
@@ -204,26 +211,21 @@
-             * we need to prompt the user to enter a little more information
-             * to complete the registration process.
+            /* If the user successfully authenticated with a third-party provider, but they haven't
+             * linked the account to an Open edX account, instruct the user on how to link the accounts.
```
Nit: `linked the account to an Open edX account` -> `linked the accounts`

Rationale: It may be unclear what an `Open edX account` is - it may also mean something else in the future.

robrap replied:
Updated.

nasthagiri commented on openedx/core/djangoapps/user_authn/views/tests/test_login.py:
```diff
@@ -666,10 +666,14 @@ def test_login_form(self):
         self.assertHttpOK(response)
 
+        # TODO: ARCH-1253: LoginSession GET returns a form that temporarily points to a different login URL for POST
+        # until we complete the roll-out/transition to the login_user response without the shim.
```
Nit: Explain what we are doing in the meantime `until we ...`.
For example: `Use the POST login_api endpoint until we ...`

robrap replied:
Updated.

nasthagiri commented on lms/static/js/student_account/views/LoginView.js:
```diff
@@ -204,26 +211,21 @@
-                if (error.status === 403 &&
-                 error.responseText === 'third-party-auth' &&
-                 this.currentProvider) {
+                if (errorCode === 'third-party-auth-with-no-linked-account' && this.currentProvider) {
```
Question: Since I haven't been following your recent PRs, I'm unaware of this, and I'm sure you covered it in another PR.  So, has `third-party-auth` changed to `third-party-auth-with-no-linked-account`?

robrap replied:
I did it when I introduced JSON for this error in this recent PR: https://github.com/edx/edx-platform/pull/22452/files

The shim currently is blowing that JSON away and using the earlier `third-party-auth`, but that seemed too vague to me for the error code.

robrap commented on lms/static/js/student_account/views/LoginView.js:
```diff
@@ -189,11 +189,18 @@
-                var msg = error.responseText;
+                var errorCode;
+                var msg;
+                // TODO: Add back the following failsafe line once tests are passing,
```
Note: This code has been updated.

robrap commented:
@nasthagiri @dianakhuang: I added two new commits to this approved PR because when working on removing the shim code, its comments made me more paranoid.  See https://github.com/edx/edx-platform/blob/f4f070bda8be69dc77c5b46f84bc0eb56333c77e/openedx/core/djangoapps/user_authn/views/login.py#L555-L574
1. The first commit adds a Django Setting so I can enable/disable the switch to the shimless api.  I think third-party auth success will be fine, but I am not certain.
2. I added custom metrics to see if they might help get to the bottom of a little mystery I have around an earlier change I made.  @dianakhuang looked at this a little with me.
